docs/okeps/okep-4456-egressip-on-user-defined-networks.md

```diff
@@ -0,0 +1,168 @@
+
+When traffic from both EgressIP and Non-EgressIP pods arrive at br-ex with GR masquerade IP used as the source IP address, there should be a way to distinguish the packet belongs which pod so that SNAT for EgressIP can applied, otherwise it would be SNATed with node's IP address. The EgressIP pod is packet marked using OVN's Logical Router Policy (LRP) and flow rule in br-ex match on both GR IP address and `pkt_mark` to apply SNAT.
+
+There is a need for each EgressIP object mapped with a unique packet mark identifier so that it doesn't create conflict between two different EgressIP objects serving two different namespaces but managed by the same user defined network. Since the name of the EgressIP object is a unique value globally, This can be used to derive packet marker identifiers using go libraries like `hash/fnv`, The maximum allowed value for the identifier is 2^32-1.
+
+### L3 user defined primary network
+
+The existing default primary network implementation for EgressIP is enhanced with following.
+
+* The pod node's ovn-cluster-router LRP entry must be updated with `pkt_mark`. The `pkt_mark` doesn't survive if the packet is put on the wire, so LRP entry must be
+also created on remote nodes to do packet marking if EgressIP is assigned there.
+* The programming of GR SNAT entries must be skipped.
+* Create appropriate flow rules in `Table 0` matching on `pkt_mark` with actions `snat(<EIP>, zone 0, CT Mark 5, table=0)`. This would need flow entry per EgressIP object and not at per pod level. So this flow rule can just exist as long as EgressIP is assigned to the node.
+
+These changes would be applicable for both shared and local gateway modes.
+
+The following section provides how EgressIP flow is handled in L3 user defined network scenarios.
+
+#### podA on nodeA and served by eip on nodA
+
+![single node](images/eip-single-node-shared.svg)
+
+#### podA on nodeA served by eip on nodeB
+
+![remote node](images/eip-remote-node-shared.svg)
+
+#### podA on nodeA served by two EIPs from two different objects one on nodeA and other on nodeB
+
+![two EIP objects for same pod](images/eips-for-same-pod-shared.svg)
+
+#### podA on nodeA served by twp EIPs from same object where one is nodeB and other is nodeC
+
+![two EIP objects for same pod remote nodes](images/eips-for-same-pod-remote-nodes-shared.svg)
+
+### L2 user defined primary network
+
+For layer 2, egress IP has never been supported before. With the IC design, there is no need to have an ovn_cluster_router and join switch separating the layer 2 switch network (transit switch) from the GR.
+
+To handle this scenario, the corresponding LRP must be created in GR to do packet marking and SNAT must happen in br-ex as similar to L3 user defined networks.
```

Comment:
The GR must also handle redirects via GW LRPs when a node contains a pod thats selected by an EIP.